lib/base/type.cpp

```diff
@@ -32,6 +38,60 @@ INITIALIZE_ONCE_WITH_PRIORITY([]() {
 	Type::Register(type);
 }, InitializePriority::RegisterTypeType);
 
+static std::vector<Type::Ptr> l_SortedByLoadDependencies;
+static Atomic l_SortingByLoadDependenciesDone (false);
+
+typedef std::unordered_map<Type*, bool> Visited; // https://stackoverflow.com/a/8942986
+
+INITIALIZE_ONCE_WITH_PRIORITY([] {
+	auto types (Type::GetAllTypes());
+
+	types.erase(std::remove_if(types.begin(), types.end(), [](auto& type) {
+		return !ConfigObject::TypeInstance->IsAssignableFrom(type);
+	}), types.end());
+
+	// Depth-first search
+	std::unordered_set<Type*> unsorted;
+	Visited visited;
+	std::vector<Type::Ptr> sorted;
+
+	for (auto type : types) {
+		unsorted.emplace(type.get());
+	}
+
+	std::function<void(Type*)> visit;
+	visit = ([&visit, &unsorted, &visited, &sorted](Type* type) {
+		if (unsorted.find(type) == unsorted.end()) {
+			return;
+		}
+
+		bool& alreadyVisited (visited.at(type));
+		VERIFY(!alreadyVisited);
```

Comment:
Is that `VERIFY()` cycle detection in disguise? It would be more helpful if it actually said so as it would give you more of a clue of what happened if you manage to trigger it.